Upgrades overhaul part 4c: Cargo and Passenger

pchote opened the pull request:
Depends on #12406.

The `GrantUpgrades` field on `Passenger` has been moved to `PassengerConditions` on `Cargo`.  This avoids adding the (IMO unnecessary) complexity of external conditions to the cargo code.

Requiring the conditions to be defined for every actor type might be controversial, so an alternative could be to rework the cargo code to support multiple `CargoType`s on a `Passenger`, and then have the conditions keyed on this.

The test case extends the one from #12406 to show how passenger conditions can be stacked for improved bonuses (2x e1 = green, 3x e1 = blue).

pchote committed:
Convert Cargo to conditions.

pchote commented:
Rebased and ready for review.

GraionDilach commented:
TBH you'd use the passenger upgrades for two reasons:
 - have additional logics based on passenger inside (IFV)
 - have a storage indicator besides the pips

This PR shifts the engine preferability - I dunno how to say it else - from the latter to the former. I'd advise a simple `Cargo`>`LoadedCondition` (with each passenger granting an instance) along with the proposed passenger-specific conditions list, sounds simple enough to me.

obrakmann commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -56,8 +56,12 @@ public class CargoInfo : ITraitInfo, Requires<IOccupySpaceInfo>
+		public readonly string LoadingCondition = null;
+
+		[UpgradeGrantedReference]
+		[Desc("Conditions to grant when specified actors are loaded inside the transport.")]
```
Please explain that the first string should be an actor type and the second the condition.

obrakmann commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -555,6 +555,19 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20161204)
```
Don't forget to update this eventually.

obrakmann reviewed (approved):
Looks otherwise fine to me, 👍

pchote committed:
Add Cargo.LoadedCondition.

pchote commented:
Updated.  I have moved the testcases to separate branches: `cargo-conditions-passengertest` for the `PassengerConditions` and `cargo-conditions-loadedtest` for the new `LoadedCondition`.

pchote commented:
Rebased (again...).

abcdefg30 commented:
👍 / ✅

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6894f77f04517a6684ad581fdea1a8fbe92fe7cf...34b23523a2629ea57e271546ec7f2d7686436874)
